Bluetooth: controller: Fix encryption issues regarding rejection and termination

thoh-ot opened the pull request:
EBQ test run passes (https://github.com/zephyrproject-rtos/zephyr/issues/43435):
- LL/SEC/PER/BI-04-C
- LL/SEC/PER/BI-03-C
- LL/SEC/PER/BI-01-C
- LL/SEC/CEN/BV-12-C
- LL/SEC/CEN/BV-13-C
- LL/SEC/CEN/BV-03-C
- LL/SEC/CEN/BV-14-C
- LL/SEC/PER/BI-05-C
- LL/SEC/PER/BV-02-C
- LL/SEC/PER/BV-01-C

Builds on top of the fixes in
https://github.com/zephyrproject-rtos/zephyr/pull/44275
https://github.com/zephyrproject-rtos/zephyr/pull/44014
https://github.com/zephyrproject-rtos/zephyr/pull/44003

thoh-ot committed:
Bluetooth: controller: Fix enc APPTO

The logic condition for when to enable APPTO needs to look at both
enc_rx and enc_tx.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Fix periph enc reject

Chose correct rejection PDU based on features supported on remote peer
when rejecting due to missing LTK.

Update unit test setup with faking that a feature exchange procedure
has run.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_enc.c:
```diff
@@ -406,6 +406,14 @@ static void lp_enc_st_wait_rx_enc_rsp(struct ll_conn *conn, struct proc_ctx *ctx
+		 * only PDUs required by this procedure are expected;
+		 * to achieve that even "unwanted" PDU's end up in this FSM,
+		 * enable the greedy RX behaviour.
+		 */
```
Suggestion for reformulation:
	/* After the Central has received the LL_ENC_RSP PDU,
		 * only PDUs related to this procedure are valid, and invalids should 
		 * result in disconnect.
		 * to achieve this enable the greedy RX behaviour, such that
		 * all PDU's end up in this FSM. 
		 */

thoh-ot replied:
ACK

erbr-ot reviewed (dismissed):
Looks real solid to me. Only a minor comment re. clarification in a comment

thoh-ot commented:
Will wait with the rebase until https://github.com/zephyrproject-rtos/zephyr/pull/44275 is merged

thoh-ot commented:
Rebased

thoh-ot committed:
Bluetooth: controller: Fix periph enc termination

Terminate connection with a MIC failure if an unexpected control PDU
is received during the Encryption Start procedure.

Add unit test inspired by Bluetooth Qualification test
LL/SEC/PER/BI-05-C,
Peripheral Receiving unexpected PDU during encryption start

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Fix central enc reject

Add support for both LL_REJECT_IND and LL_REJECT_EXT_IND when waiting
for the the response to the LL_ENC_REQ and LL_START_ENC_REQ.

Add unit test to test both LL_REJECT_IND and LL_REJECT_EXT_IND as
responses to LL_ENC_REQ.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Fix central enc termination

Terminate connection with a MIC failure if an unexpected control PDU
is received during the Encryption Start procedure.

Add a greedy option to pdu_is_expected() to make sure the procedure
processes all unexpected control PDU in all cases.

Add unit test inspired by Bluetooth Qualification test
LL/SEC/CEN/BV-14-C,
Central Receiving unexpected PDU during encryption start

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

erbr-ot reviewed (approved):
LGTM
